<Type Name="NullReferenceException" FullName="System.NullReferenceException">
  <Metadata><Meta Name="ms.openlocfilehash" Value="a0c06062ed05c53266ca301808bbaf02d6034e28" /><Meta Name="ms.sourcegitcommit" Value="055a4a82a0b08bfbdc21bd1347fb71f7fe2c099e" /><Meta Name="ms.translationtype" Value="MT" /><Meta Name="ms.contentlocale" Value="ru-RU" /><Meta Name="ms.lasthandoff" Value="08/15/2019" /><Meta Name="ms.locfileid" Value="69419755" /></Metadata><TypeSignature Language="C#" Value="public class NullReferenceException : SystemException" />
  <TypeSignature Language="ILAsm" Value=".class public auto ansi serializable beforefieldinit NullReferenceException extends System.SystemException" />
  <TypeSignature Language="DocId" Value="T:System.NullReferenceException" />
  <TypeSignature Language="VB.NET" Value="Public Class NullReferenceException&#xA;Inherits SystemException" />
  <TypeSignature Language="C++ CLI" Value="public ref class NullReferenceException : SystemException" />
  <TypeSignature Language="F#" Value="type NullReferenceException = class&#xA;    inherit SystemException" />
  <AssemblyInfo>
    <AssemblyName>System.Runtime</AssemblyName>
    <AssemblyVersion>4.0.0.0</AssemblyVersion>
    <AssemblyVersion>4.0.10.0</AssemblyVersion>
    <AssemblyVersion>4.0.20.0</AssemblyVersion>
    <AssemblyVersion>4.1.0.0</AssemblyVersion>
    <AssemblyVersion>4.2.0.0</AssemblyVersion>
    <AssemblyVersion>4.2.1.0</AssemblyVersion>
  </AssemblyInfo>
  <AssemblyInfo>
    <AssemblyName>mscorlib</AssemblyName>
    <AssemblyVersion>1.0.5000.0</AssemblyVersion>
    <AssemblyVersion>2.0.0.0</AssemblyVersion>
    <AssemblyVersion>2.0.5.0</AssemblyVersion>
    <AssemblyVersion>4.0.0.0</AssemblyVersion>
  </AssemblyInfo>
  <AssemblyInfo>
    <AssemblyName>netstandard</AssemblyName>
    <AssemblyVersion>2.0.0.0</AssemblyVersion>
    <AssemblyVersion>2.1.0.0</AssemblyVersion>
  </AssemblyInfo>
  <Base>
    <BaseTypeName>System.Exception</BaseTypeName>
    <BaseTypeName FrameworkAlternate="netframework-4.5">System.SystemException</BaseTypeName>
  </Base>
  <Interfaces />
  <Attributes>
    <Attribute FrameworkAlternate="netframework-2.0;netframework-3.0;netframework-3.5;netframework-4.0;netframework-4.5;netframework-4.5.1;netframework-4.5.2;netframework-4.6;netframework-4.6.1;netframework-4.6.2;netframework-4.7;netframework-4.7.1;netframework-4.7.2;xamarinandroid-7.1;xamarinios-10.8;xamarinmac-3.0;netframework-4.8">
      <AttributeName>System.Runtime.InteropServices.ComVisible(true)</AttributeName>
    </Attribute>
    <Attribute FrameworkAlternate="netframework-1.1;netframework-2.0;netframework-3.0;netframework-3.5;netframework-4.0;netframework-4.5;netframework-4.5.1;netframework-4.5.2;netframework-4.6;netframework-4.6.1;netframework-4.6.2;netframework-4.7;netframework-4.7.1;netframework-4.7.2;netframework-4.8;xamarinandroid-7.1;xamarinios-10.8;xamarinmac-3.0">
      <AttributeName>System.Serializable</AttributeName>
    </Attribute>
  </Attributes>
  <Docs>
    <summary><span data-ttu-id="f05e3-101">Исключение, возникающее при попытке разыменования указателя NULL на объект.</span><span class="sxs-lookup"><span data-stu-id="f05e3-101">The exception that is thrown when there is an attempt to dereference a null object reference.</span></span></summary>
    <remarks>
      <format type="text/markdown"><![CDATA[  
  
## Remarks  
 <span data-ttu-id="f05e3-102">Исключение возникает при попытке получить доступ к члену для типа, значение которого равно `null`. <xref:System.NullReferenceException></span><span class="sxs-lookup"><span data-stu-id="f05e3-102">A <xref:System.NullReferenceException> exception is thrown when you try to access a member on a type whose value is `null`.</span></span> <span data-ttu-id="f05e3-103"><xref:System.NullReferenceException> Исключение обычно отражает ошибку разработчика и возникает в следующих сценариях:</span><span class="sxs-lookup"><span data-stu-id="f05e3-103">A <xref:System.NullReferenceException> exception typically reflects developer error and is thrown in the following scenarios:</span></span>  
  
-   <span data-ttu-id="f05e3-104">Вы забыли создать экземпляр ссылочного типа.</span><span class="sxs-lookup"><span data-stu-id="f05e3-104">You've forgotten to instantiate a reference type.</span></span> <span data-ttu-id="f05e3-105">В следующем примере объявляется `names` , но никогда не создается экземпляр:</span><span class="sxs-lookup"><span data-stu-id="f05e3-105">In the following example, `names` is declared but never instantiated:</span></span>  
  
     [!code-csharp[System.NullReferenceException.Class#1](~/samples/snippets/csharp/VS_Snippets_CLR_System/system.nullreferenceexception.class/cs/example1.cs#1)]
     [!code-vb[System.NullReferenceException.Class#1](~/samples/snippets/visualbasic/VS_Snippets_CLR_System/system.nullreferenceexception.class/vb/example1.vb#1)]  
  
     <span data-ttu-id="f05e3-106">Некоторые компиляторы выдают предупреждение при компиляции этого кода.</span><span class="sxs-lookup"><span data-stu-id="f05e3-106">Some compilers issue a warning when they compile this code.</span></span> <span data-ttu-id="f05e3-107">Другие выдают ошибку, и компиляция завершается ошибкой.</span><span class="sxs-lookup"><span data-stu-id="f05e3-107">Others issue an error, and the compilation fails.</span></span> <span data-ttu-id="f05e3-108">Чтобы устранить эту проблему, создайте экземпляр объекта, чтобы его значение больше не `null`было.</span><span class="sxs-lookup"><span data-stu-id="f05e3-108">To address this problem, instantiate the object so that its value is no longer `null`.</span></span> <span data-ttu-id="f05e3-109">В следующем примере это делается путем вызова конструктора класса типа.</span><span class="sxs-lookup"><span data-stu-id="f05e3-109">The following example does this by calling a type's class constructor.</span></span>  
  
     [!code-csharp[System.NullReferenceException.Class#2](~/samples/snippets/csharp/VS_Snippets_CLR_System/system.nullreferenceexception.class/cs/example1a.cs#2)]
     [!code-vb[System.NullReferenceException.Class#2](~/samples/snippets/visualbasic/VS_Snippets_CLR_System/system.nullreferenceexception.class/vb/example1a.vb#2)]  
  
-   <span data-ttu-id="f05e3-110">Вы забыли размер массива перед его инициализацией.</span><span class="sxs-lookup"><span data-stu-id="f05e3-110">You've forgotten to dimension an array before initializing it.</span></span> <span data-ttu-id="f05e3-111">В следующем примере `values` объявляется как массив целых чисел, но число содержащихся в нем элементов никогда не указывается.</span><span class="sxs-lookup"><span data-stu-id="f05e3-111">In the following example, `values` is declared to be an integer array, but the number of elements that it contains is never specified.</span></span> <span data-ttu-id="f05e3-112">Таким образом, попытка инициализировать свои значения вызвала <xref:System.NullReferenceException> исключение.</span><span class="sxs-lookup"><span data-stu-id="f05e3-112">The attempt to initialize its values therefore thrown a  <xref:System.NullReferenceException> exception.</span></span>  
  
     [!code-csharp[System.NullReferenceException.Class#10](~/samples/snippets/csharp/VS_Snippets_CLR_System/system.nullreferenceexception.class/cs/Array3.cs#10)]
     [!code-vb[System.NullReferenceException.Class#10](~/samples/snippets/visualbasic/VS_Snippets_CLR_System/system.nullreferenceexception.class/vb/Array3.vb#10)]  
  
     <span data-ttu-id="f05e3-113">Исключение можно устранить, объявив число элементов в массиве перед его инициализацией, как показано в следующем примере.</span><span class="sxs-lookup"><span data-stu-id="f05e3-113">You can eliminate the exception by declaring the number of elements in the array before initializing it, as the following example does.</span></span>  
  
     [!code-csharp[System.NullReferenceException.Class#11](~/samples/snippets/csharp/VS_Snippets_CLR_System/system.nullreferenceexception.class/cs/Array4.cs#11)]
     [!code-vb[System.NullReferenceException.Class#11](~/samples/snippets/visualbasic/VS_Snippets_CLR_System/system.nullreferenceexception.class/vb/Array4.vb#11)]  
  
     <span data-ttu-id="f05e3-114">Дополнительные сведения об объявлении и инициализации массивов см. [](~/docs/csharp/programming-guide/arrays/index.md) в разделе массивы и [массивы](~/docs/visual-basic/programming-guide/language-features/arrays/index.md).</span><span class="sxs-lookup"><span data-stu-id="f05e3-114">For more information on declaring and initializing arrays, see [Arrays](~/docs/csharp/programming-guide/arrays/index.md) and [Arrays](~/docs/visual-basic/programming-guide/language-features/arrays/index.md).</span></span>  
  
-   <span data-ttu-id="f05e3-115">Вы получаете возвращаемое значение **null** из метода, а затем вызываете метод для возвращаемого типа.</span><span class="sxs-lookup"><span data-stu-id="f05e3-115">You get a **null** return value from a method, and then call a method on the returned type.</span></span> <span data-ttu-id="f05e3-116">Это иногда приводит к ошибке в документации. в документации не заметка о том, что вызов метода `null`может вернуть.</span><span class="sxs-lookup"><span data-stu-id="f05e3-116">This sometimes is the result of a documentation error; the documentation fails to note that a method call can return `null`.</span></span> <span data-ttu-id="f05e3-117">В других случаях код ошибочно предполагает, что метод всегда будет возвращать значение, отличное от**null** .</span><span class="sxs-lookup"><span data-stu-id="f05e3-117">In other cases, your code erroneously assumes that the method will always return a non-**null** value.</span></span>  
  
     <span data-ttu-id="f05e3-118">Код в следующем примере предполагает, что <xref:System.Array.Find%2A?displayProperty=nameWithType> метод всегда возвращает `Person` объект, поле которого `FirstName` соответствует строке поиска.</span><span class="sxs-lookup"><span data-stu-id="f05e3-118">The code in the following example assumes that the <xref:System.Array.Find%2A?displayProperty=nameWithType> method always returns `Person` object whose `FirstName` field matches a search string.</span></span> <span data-ttu-id="f05e3-119">Поскольку совпадений нет, среда выполнения создает <xref:System.NullReferenceException> исключение.</span><span class="sxs-lookup"><span data-stu-id="f05e3-119">Because there is no match, the runtime throws a <xref:System.NullReferenceException> exception.</span></span>  
  
     [!code-csharp[System.NullReferenceException.Class#4](~/samples/snippets/csharp/VS_Snippets_CLR_System/system.nullreferenceexception.class/cs/nullreturn2.cs#4)]
     [!code-vb[System.NullReferenceException.Class#4](~/samples/snippets/visualbasic/VS_Snippets_CLR_System/system.nullreferenceexception.class/vb/nullreturn2.vb#4)]  
  
     <span data-ttu-id="f05e3-120">Чтобы устранить эту проблему, проверьте возвращаемое значение метода, чтобы убедиться, что он не `null` перед вызовом любого из его членов, как показано в следующем примере.</span><span class="sxs-lookup"><span data-stu-id="f05e3-120">To address this problem, test the method's return value to ensure that it is not `null` before calling any of its members, as the following example does.</span></span>  
  
     [!code-csharp[System.NullReferenceException.Class#5](~/samples/snippets/csharp/VS_Snippets_CLR_System/system.nullreferenceexception.class/cs/nullreturn2a.cs#5)]
     [!code-vb[System.NullReferenceException.Class#5](~/samples/snippets/visualbasic/VS_Snippets_CLR_System/system.nullreferenceexception.class/vb/nullreturn2a.vb#5)]  
  
-   <span data-ttu-id="f05e3-121">Вы используете выражение (например, объединяете список методов или свойств) для получения значения и, несмотря на то, что проверяется, является `null`ли значение, среда выполнения по-прежнему <xref:System.NullReferenceException> вызывает исключение.</span><span class="sxs-lookup"><span data-stu-id="f05e3-121">You're using an expression (for example, you're chaining a list of methods or properties together) to retrieve a value and, although you're checking whether the value is `null`, the runtime still throws a <xref:System.NullReferenceException> exception.</span></span> <span data-ttu-id="f05e3-122">Это происходит из-за того, что одно из промежуточных `null`значений в выражении возвращает.</span><span class="sxs-lookup"><span data-stu-id="f05e3-122">This occurs because one of the intermediate values in the expression returns `null`.</span></span> <span data-ttu-id="f05e3-123">В результате тест для `null` никогда не вычисляется.</span><span class="sxs-lookup"><span data-stu-id="f05e3-123">As a result, your test for `null` is never evaluated.</span></span>  
  
     <span data-ttu-id="f05e3-124">В следующем примере определяется `Pages` объект, который кэширует сведения о веб-страницах, которые `Page` представлены объектами.</span><span class="sxs-lookup"><span data-stu-id="f05e3-124">The following example defines a `Pages` object that  caches information about web pages, which are presented by `Page` objects.</span></span> <span data-ttu-id="f05e3-125">`Example.Main` Метод проверяет, имеет ли текущая веб-страница заголовок, отличный от NULL, и отображает заголовок.</span><span class="sxs-lookup"><span data-stu-id="f05e3-125">The `Example.Main` method checks whether the current web page has a non-null title and, if it does, displays the title.</span></span> <span data-ttu-id="f05e3-126">Однако несмотря на эту проверку, метод создает <xref:System.NullReferenceException> исключение.</span><span class="sxs-lookup"><span data-stu-id="f05e3-126">Despite this check, however, the method throws a <xref:System.NullReferenceException> exception.</span></span>  
  
     [!code-csharp[System.NullReferenceException.Class#6](~/samples/snippets/csharp/VS_Snippets_CLR_System/system.nullreferenceexception.class/cs/Chain1.cs#6)]
     [!code-vb[System.NullReferenceException.Class#6](~/samples/snippets/visualbasic/VS_Snippets_CLR_System/system.nullreferenceexception.class/vb/Chain1.vb#6)]  
  
     <span data-ttu-id="f05e3-127">Исключение возникает, поскольку `pages.CurrentPage` возвращает `null` , если сведения о странице не хранятся в кэше.</span><span class="sxs-lookup"><span data-stu-id="f05e3-127">The exception is thrown because `pages.CurrentPage` returns `null` if no page information is  stored in the cache.</span></span> <span data-ttu-id="f05e3-128">Это исключение можно исправить, проверив значение `CurrentPage` свойства перед получением `Title` свойства текущего `Page` объекта, как показано в следующем примере:</span><span class="sxs-lookup"><span data-stu-id="f05e3-128">This exception can be corrected by testing the value of the `CurrentPage` property before retrieving the current `Page` object's `Title` property, as the following example does:</span></span>  
  
     [!code-csharp[System.NullReferenceException.Class#7](~/samples/snippets/csharp/VS_Snippets_CLR_System/system.nullreferenceexception.class/cs/Chain2.cs#7)]
     [!code-vb[System.NullReferenceException.Class#7](~/samples/snippets/visualbasic/VS_Snippets_CLR_System/system.nullreferenceexception.class/vb/Chain2.vb#7)]  
  
-   <span data-ttu-id="f05e3-129">Выполняется перечисление элементов массива, содержащего ссылочные типы, и попытка обработать один из элементов вызывает <xref:System.NullReferenceException> исключение.</span><span class="sxs-lookup"><span data-stu-id="f05e3-129">You're enumerating the elements of an array that contains reference types, and your attempt to process one of the elements throws a <xref:System.NullReferenceException> exception.</span></span>  
  
     <span data-ttu-id="f05e3-130">В следующем примере определяется массив строк.</span><span class="sxs-lookup"><span data-stu-id="f05e3-130">The following example defines a string array.</span></span> <span data-ttu-id="f05e3-131">Оператор перечисляет элементы в массиве и вызывает <xref:System.String.Trim%2A> метод каждой строки перед отображением строки. `for`</span><span class="sxs-lookup"><span data-stu-id="f05e3-131">A `for` statement enumerates the elements in the array and calls each string's <xref:System.String.Trim%2A> method before displaying the string.</span></span>  
  
     [!code-csharp[System.NullReferenceException.Class#8](~/samples/snippets/csharp/VS_Snippets_CLR_System/system.nullreferenceexception.class/cs/Array1.cs#8)]
     [!code-vb[System.NullReferenceException.Class#8](~/samples/snippets/visualbasic/VS_Snippets_CLR_System/system.nullreferenceexception.class/vb/Array1.vb#8)]  
  
     <span data-ttu-id="f05e3-132">Это исключение возникает, если предполагается, что каждый элемент массива должен содержать значение, отличное от NULL, и значение элемента массива фактически `null`.</span><span class="sxs-lookup"><span data-stu-id="f05e3-132">This exception occurs if you assume that each element of the array must contain a non-null value, and the value of the array element is in fact `null`.</span></span> <span data-ttu-id="f05e3-133">Исключение можно устранить, проверив, находится ли элемент перед `null` выполнением какой-либо операции с этим элементом, как показано в следующем примере.</span><span class="sxs-lookup"><span data-stu-id="f05e3-133">The exception can be eliminated by testing whether the element is `null` before performing any operation on that element, as the following example shows.</span></span>  
  
     [!code-csharp[System.NullReferenceException.Class#9](~/samples/snippets/csharp/VS_Snippets_CLR_System/system.nullreferenceexception.class/cs/Array2.cs#9)]
     [!code-vb[System.NullReferenceException.Class#9](~/samples/snippets/visualbasic/VS_Snippets_CLR_System/system.nullreferenceexception.class/vb/Array2.vb#9)]  
  
-   <span data-ttu-id="f05e3-134">Исключение вызывается методом, который передается `null`. <xref:System.NullReferenceException></span><span class="sxs-lookup"><span data-stu-id="f05e3-134">A <xref:System.NullReferenceException> exception is thrown by a method that is passed `null`.</span></span> <span data-ttu-id="f05e3-135">Некоторые методы проверяют передаваемые им аргументы.</span><span class="sxs-lookup"><span data-stu-id="f05e3-135">Some methods validate the arguments that are passed to them.</span></span> <span data-ttu-id="f05e3-136">Если они выполняют и один из аргументов имеет значение `null`, метод <xref:System.ArgumentNullException?displayProperty=nameWithType> создает исключение.</span><span class="sxs-lookup"><span data-stu-id="f05e3-136">If they do and one of the arguments is `null`, the method throws an <xref:System.ArgumentNullException?displayProperty=nameWithType> exception.</span></span> <span data-ttu-id="f05e3-137">В противном случае вызывается <xref:System.NullReferenceException> исключение.</span><span class="sxs-lookup"><span data-stu-id="f05e3-137">Otherwise, it throws a <xref:System.NullReferenceException> exception.</span></span> <span data-ttu-id="f05e3-138">Этот сценарий показан в следующем примере.</span><span class="sxs-lookup"><span data-stu-id="f05e3-138">The following example illustrates this scenario.</span></span>  
  
     [!code-csharp[System.NullReferenceException.Class#3](~/samples/snippets/csharp/VS_Snippets_CLR_System/system.nullreferenceexception.class/cs/example2.cs#3)]
     [!code-vb[System.NullReferenceException.Class#3](~/samples/snippets/visualbasic/VS_Snippets_CLR_System/system.nullreferenceexception.class/vb/example2.vb#3)]  
  
     <span data-ttu-id="f05e3-139">Чтобы устранить эту ошибку, убедитесь, что аргумент, передаваемый в метод, не `null`имеет значение, или обрабатывайте созданное `try…catch…finally` исключение в блоке.</span><span class="sxs-lookup"><span data-stu-id="f05e3-139">To address this issue, make sure that the argument passed to the method is not `null`, or handle the thrown exception in a `try…catch…finally` block.</span></span> <span data-ttu-id="f05e3-140">Дополнительные сведения см. в разделе [Исключения](~/docs/standard/exceptions/index.md).</span><span class="sxs-lookup"><span data-stu-id="f05e3-140">For more information, see [Exceptions](~/docs/standard/exceptions/index.md).</span></span>  
  
 <span data-ttu-id="f05e3-141">Следующие инструкции <xref:System.NullReferenceException>ПРОМЕЖУТОЧного языка MSIL вызовут: `ldelema` `ldelem.<type>` `cpobj` `callvirt`, `cpblk`,, `initblk`,,, `ldfld`, ,`ldflda` `ldind.<type>` , `ldlen` ,`stelem.<type>`, ,`stind.<type>`, и`throw`. `stfld` `unbox`</span><span class="sxs-lookup"><span data-stu-id="f05e3-141">The following Microsoft intermediate language (MSIL) instructions throw <xref:System.NullReferenceException>: `callvirt`, `cpblk`, `cpobj`, `initblk`, `ldelem.<type>`, `ldelema`, `ldfld`, `ldflda`, `ldind.<type>`, `ldlen`, `stelem.<type>`, `stfld`, `stind.<type>`, `throw`, and `unbox`.</span></span>  
  
 <span data-ttu-id="f05e3-142"><xref:System.NullReferenceException>использует HRESULT COR_E_NULLREFERENCE, имеющий значение 0x80004003.</span><span class="sxs-lookup"><span data-stu-id="f05e3-142"><xref:System.NullReferenceException> uses the HRESULT COR_E_NULLREFERENCE, which has the value 0x80004003.</span></span>  
  
 <span data-ttu-id="f05e3-143">Список начальных значений свойств для экземпляра <xref:System.NullReferenceException>, см. в разделе <xref:System.NullReferenceException.%23ctor%2A> конструкторы.</span><span class="sxs-lookup"><span data-stu-id="f05e3-143">For a list of initial property values for an instance of <xref:System.NullReferenceException>, see the <xref:System.NullReferenceException.%23ctor%2A> constructors.</span></span>  
  
 <span data-ttu-id="f05e3-144">**Обработка NullReferenceException в коде выпуска**</span><span class="sxs-lookup"><span data-stu-id="f05e3-144">**Handling NullReferenceException in release code**</span></span>  
  
 <span data-ttu-id="f05e3-145">Обычно лучше избегать NullReferenceException, чем после появления.</span><span class="sxs-lookup"><span data-stu-id="f05e3-145">It's usually better to avoid a NullReferenceException than to handle it after it occurs.</span></span> <span data-ttu-id="f05e3-146">Обработка исключений может затруднять обслуживание и понимание вашего кода, а иногда даже приводить к другим ошибкам.</span><span class="sxs-lookup"><span data-stu-id="f05e3-146">Handling an exception can make your code harder to maintain and understand, and can sometimes introduce other bugs.</span></span> <span data-ttu-id="f05e3-147">Исключение NullReferenceException часто является неисправимой ошибкой.</span><span class="sxs-lookup"><span data-stu-id="f05e3-147">A NullReferenceException is often a non-recoverable error.</span></span> <span data-ttu-id="f05e3-148">В этом случае, возможно, будет лучше позволить исключению остановить приложение.</span><span class="sxs-lookup"><span data-stu-id="f05e3-148">In these cases, letting the exception stop the app might be the best alternative.</span></span>  
  
 <span data-ttu-id="f05e3-149">Однако существует много ситуаций, в которых процесс обработки ошибки может оказаться полезен:</span><span class="sxs-lookup"><span data-stu-id="f05e3-149">However, there are many situations where handling the error can be useful:</span></span>  
  
-   <span data-ttu-id="f05e3-150">Приложение может пропускать объекты, которые имеют значение null.</span><span class="sxs-lookup"><span data-stu-id="f05e3-150">Your app can ignore objects that are null.</span></span> <span data-ttu-id="f05e3-151">Например, если приложение получает и обрабатывает записи в базе данных, можно игнорировать некоторое число неверных записей, которые вызывают возникновение объектов со значением null.</span><span class="sxs-lookup"><span data-stu-id="f05e3-151">For example, if your app retrieves and processes records in a database, you might be able to ignore some number of bad records that result in null objects.</span></span> <span data-ttu-id="f05e3-152">Возможно, все, что потребуется, — записать неверные данные в файл журнала или интерфейс пользователя приложения.</span><span class="sxs-lookup"><span data-stu-id="f05e3-152">Recording the bad data in a log file or in the application UI might be all you have to do.</span></span>  
  
-   <span data-ttu-id="f05e3-153">Можно попытаться исправить исключение.</span><span class="sxs-lookup"><span data-stu-id="f05e3-153">You can recover from the exception.</span></span> <span data-ttu-id="f05e3-154">Например, вызов веб-службы, который возвращает ссылочный тип, может вернуть значение null в случае разрыва или истечения времени ожидания подключения. Можно попытаться восстановить подключение и повторить вызов.</span><span class="sxs-lookup"><span data-stu-id="f05e3-154">For example, a call to a web service that returns a reference type might return null if the connection is lost or the connection times out. You can attempt to reestablish the connection and try the call again.</span></span>  
  
-   <span data-ttu-id="f05e3-155">Можно восстановить приложение до допустимого состояния.</span><span class="sxs-lookup"><span data-stu-id="f05e3-155">You can restore the state of your app to a valid state.</span></span> <span data-ttu-id="f05e3-156">Например, вы могли выполнять задачу, состоящую из нескольких шагов, которая требует сохранения информации в хранилище данных, перед вызовом метода, который создает исключение NullReferenceException.</span><span class="sxs-lookup"><span data-stu-id="f05e3-156">For example, you might be performing a multi-step task that requires you to save information to a data store before you call a method that throws a NullReferenceException.</span></span> <span data-ttu-id="f05e3-157">Если неинициализированный объект повредит запись данных, можно будет удалить предыдущие данные перед закрытием приложения.</span><span class="sxs-lookup"><span data-stu-id="f05e3-157">If the uninitialized object would corrupt the data record, you can remove the previous data before you close the app.</span></span>  
  
-   <span data-ttu-id="f05e3-158">Требуется сообщить об исключении.</span><span class="sxs-lookup"><span data-stu-id="f05e3-158">You want to report the exception.</span></span> <span data-ttu-id="f05e3-159">Например, если ошибка была вызвана неправильным действием пользователя приложения, можно создать сообщение, которое поможет ему предоставить правильные сведения.</span><span class="sxs-lookup"><span data-stu-id="f05e3-159">For example, if the error was caused by a mistake from the user of your app, you can generate a message to help him supply the correct information.</span></span> <span data-ttu-id="f05e3-160">Кроме того, можно зарегистрировать сведения об ошибке, которые помогут устранить проблему.</span><span class="sxs-lookup"><span data-stu-id="f05e3-160">You can also log information about the error to help you fix the problem.</span></span> <span data-ttu-id="f05e3-161">Некоторые платформы, например, ASP.NET, содержат высокогоуровневый обработчик исключений, который перехватывает все ошибки, чтобы приложения никогда аварийно не завершало работу; в этом случае только благодаря регистрации исключения вы сможете узнать о наличии ошибки.</span><span class="sxs-lookup"><span data-stu-id="f05e3-161">Some frameworks, like ASP.NET, have a high-level exception handler that captures all errors to that the app never crashes; in that case, logging the exception might be the only way you can know that it occurs.</span></span>  
  
 ]]></format>
    </remarks>
    <altmember cref="T:System.Exception" />
    <related type="Article" href="~/docs/standard/exceptions/index.md"><span data-ttu-id="f05e3-162">Обработка и создание исключений</span><span class="sxs-lookup"><span data-stu-id="f05e3-162">Handling and Throwing Exceptions</span></span></related>
  </Docs>
  <Members>
    <MemberGroup MemberName=".ctor">
      <AssemblyInfo>
        <AssemblyName>System.Runtime</AssemblyName>
        <AssemblyVersion>4.0.0.0</AssemblyVersion>
        <AssemblyVersion>4.0.10.0</AssemblyVersion>
        <AssemblyVersion>4.0.20.0</AssemblyVersion>
        <AssemblyVersion>4.1.0.0</AssemblyVersion>
        <AssemblyVersion>4.2.0.0</AssemblyVersion>
      </AssemblyInfo>
      <Docs>
        <summary><span data-ttu-id="f05e3-163">Инициализирует новый экземпляр класса <see cref="T:System.NullReferenceException" />.</span><span class="sxs-lookup"><span data-stu-id="f05e3-163">Initializes a new instance of the <see cref="T:System.NullReferenceException" /> class.</span></span></summary>
      </Docs>
    </MemberGroup>
    <Member MemberName=".ctor">
      <MemberSignature Language="C#" Value="public NullReferenceException ();" />
      <MemberSignature Language="ILAsm" Value=".method public hidebysig specialname rtspecialname instance void .ctor() cil managed" />
      <MemberSignature Language="DocId" Value="M:System.NullReferenceException.#ctor" />
      <MemberSignature Language="VB.NET" Value="Public Sub New ()" />
      <MemberSignature Language="C++ CLI" Value="public:&#xA; NullReferenceException();" />
      <MemberType>Constructor</MemberType>
      <AssemblyInfo>
        <AssemblyName>System.Runtime</AssemblyName>
        <AssemblyVersion>4.0.0.0</AssemblyVersion>
        <AssemblyVersion>4.0.10.0</AssemblyVersion>
        <AssemblyVersion>4.0.20.0</AssemblyVersion>
        <AssemblyVersion>4.1.0.0</AssemblyVersion>
        <AssemblyVersion>4.2.0.0</AssemblyVersion>
        <AssemblyVersion>4.2.1.0</AssemblyVersion>
      </AssemblyInfo>
      <AssemblyInfo>
        <AssemblyName>mscorlib</AssemblyName>
        <AssemblyVersion>1.0.5000.0</AssemblyVersion>
        <AssemblyVersion>2.0.0.0</AssemblyVersion>
        <AssemblyVersion>2.0.5.0</AssemblyVersion>
        <AssemblyVersion>4.0.0.0</AssemblyVersion>
      </AssemblyInfo>
      <AssemblyInfo>
        <AssemblyName>netstandard</AssemblyName>
        <AssemblyVersion>2.0.0.0</AssemblyVersion>
        <AssemblyVersion>2.1.0.0</AssemblyVersion>
      </AssemblyInfo>
      <Parameters />
      <Docs>
        <summary><span data-ttu-id="f05e3-164">Инициализирует новый экземпляр класса <see cref="T:System.NullReferenceException" />, устанавливая в качестве значения свойства <see cref="P:System.Exception.Message" /> нового экземпляра системное сообщение с описанием ошибки: "При запросе экземпляра объекта обнаружено значение NULL".</span><span class="sxs-lookup"><span data-stu-id="f05e3-164">Initializes a new instance of the <see cref="T:System.NullReferenceException" /> class, setting the <see cref="P:System.Exception.Message" /> property of the new instance to a system-supplied message that describes the error, such as "The value 'null' was found where an instance of an object was required."</span></span> <span data-ttu-id="f05e3-165">Это сообщение учитывает текущую культуру системы.</span><span class="sxs-lookup"><span data-stu-id="f05e3-165">This message takes into account the current system culture.</span></span></summary>
        <remarks>
          <format type="text/markdown"><![CDATA[  
  
## Remarks  
 <span data-ttu-id="f05e3-166">В следующей таблице представлены исходные значения свойств экземпляра класса <xref:System.NullReferenceException>.</span><span class="sxs-lookup"><span data-stu-id="f05e3-166">The following table shows the initial property values for an instance of <xref:System.NullReferenceException>.</span></span>  
  
|<span data-ttu-id="f05e3-167">Свойство.</span><span class="sxs-lookup"><span data-stu-id="f05e3-167">Property</span></span>|<span data-ttu-id="f05e3-168">Значение</span><span class="sxs-lookup"><span data-stu-id="f05e3-168">Value</span></span>|  
|--------------|-----------|  
|<xref:System.Exception.InnerException%2A>|<span data-ttu-id="f05e3-169">`null`.</span><span class="sxs-lookup"><span data-stu-id="f05e3-169">`null`.</span></span>|  
|<xref:System.Exception.Message%2A>|<span data-ttu-id="f05e3-170">Строка локализованного сообщения об ошибке.</span><span class="sxs-lookup"><span data-stu-id="f05e3-170">The localized error message string.</span></span>|  
  
 ]]></format>
        </remarks>
      </Docs>
    </Member>
    <Member MemberName=".ctor">
      <MemberSignature Language="C#" Value="public NullReferenceException (string message);" />
      <MemberSignature Language="ILAsm" Value=".method public hidebysig specialname rtspecialname instance void .ctor(string message) cil managed" />
      <MemberSignature Language="DocId" Value="M:System.NullReferenceException.#ctor(System.String)" />
      <MemberSignature Language="VB.NET" Value="Public Sub New (message As String)" />
      <MemberSignature Language="C++ CLI" Value="public:&#xA; NullReferenceException(System::String ^ message);" />
      <MemberSignature Language="F#" Value="new NullReferenceException : string -&gt; NullReferenceException" Usage="new System.NullReferenceException message" />
      <MemberType>Constructor</MemberType>
      <AssemblyInfo>
        <AssemblyName>System.Runtime</AssemblyName>
        <AssemblyVersion>4.0.0.0</AssemblyVersion>
        <AssemblyVersion>4.0.10.0</AssemblyVersion>
        <AssemblyVersion>4.0.20.0</AssemblyVersion>
        <AssemblyVersion>4.1.0.0</AssemblyVersion>
        <AssemblyVersion>4.2.0.0</AssemblyVersion>
        <AssemblyVersion>4.2.1.0</AssemblyVersion>
      </AssemblyInfo>
      <AssemblyInfo>
        <AssemblyName>mscorlib</AssemblyName>
        <AssemblyVersion>1.0.5000.0</AssemblyVersion>
        <AssemblyVersion>2.0.0.0</AssemblyVersion>
        <AssemblyVersion>2.0.5.0</AssemblyVersion>
        <AssemblyVersion>4.0.0.0</AssemblyVersion>
      </AssemblyInfo>
      <AssemblyInfo>
        <AssemblyName>netstandard</AssemblyName>
        <AssemblyVersion>2.0.0.0</AssemblyVersion>
        <AssemblyVersion>2.1.0.0</AssemblyVersion>
      </AssemblyInfo>
      <Parameters>
        <Parameter Name="message" Type="System.String" />
      </Parameters>
      <Docs>
        <param name="message"><span data-ttu-id="f05e3-171">Строка <see cref="T:System.String" />, описывающая ошибку.</span><span class="sxs-lookup"><span data-stu-id="f05e3-171">A <see cref="T:System.String" /> that describes the error.</span></span> <span data-ttu-id="f05e3-172">Содержимое параметра <paramref name="message" /> должно быть понятным пользователю.</span><span class="sxs-lookup"><span data-stu-id="f05e3-172">The content of <paramref name="message" /> is intended to be understood by humans.</span></span> <span data-ttu-id="f05e3-173">Код, вызывающий этот конструктор, должен обеспечить локализацию данной строки в соответствии с текущим языком и региональными параметрами системы.</span><span class="sxs-lookup"><span data-stu-id="f05e3-173">The caller of this constructor is required to ensure that this string has been localized for the current system culture.</span></span></param>
        <summary><span data-ttu-id="f05e3-174">Инициализирует новый экземпляр класса <see cref="T:System.NullReferenceException" /> с указанным сообщением об ошибке.</span><span class="sxs-lookup"><span data-stu-id="f05e3-174">Initializes a new instance of the <see cref="T:System.NullReferenceException" /> class with a specified error message.</span></span></summary>
        <remarks>
          <format type="text/markdown"><![CDATA[  
  
## Remarks  
 <span data-ttu-id="f05e3-175">В следующей таблице представлены исходные значения свойств экземпляра класса <xref:System.NullReferenceException>.</span><span class="sxs-lookup"><span data-stu-id="f05e3-175">The following table shows the initial property values for an instance of <xref:System.NullReferenceException>.</span></span>  
  
|<span data-ttu-id="f05e3-176">Свойство.</span><span class="sxs-lookup"><span data-stu-id="f05e3-176">Property</span></span>|<span data-ttu-id="f05e3-177">Значение</span><span class="sxs-lookup"><span data-stu-id="f05e3-177">Value</span></span>|  
|--------------|-----------|  
|<xref:System.Exception.InnerException%2A>|<span data-ttu-id="f05e3-178">`null`.</span><span class="sxs-lookup"><span data-stu-id="f05e3-178">`null`.</span></span>|  
|<xref:System.Exception.Message%2A>|<span data-ttu-id="f05e3-179">Строка сообщения об ошибке.</span><span class="sxs-lookup"><span data-stu-id="f05e3-179">The error message string.</span></span>|  
  
 ]]></format>
        </remarks>
      </Docs>
    </Member>
    <Member MemberName=".ctor">
      <MemberSignature Language="C#" Value="protected NullReferenceException (System.Runtime.Serialization.SerializationInfo info, System.Runtime.Serialization.StreamingContext context);" />
      <MemberSignature Language="ILAsm" Value=".method familyhidebysig specialname rtspecialname instance void .ctor(class System.Runtime.Serialization.SerializationInfo info, valuetype System.Runtime.Serialization.StreamingContext context) cil managed" />
      <MemberSignature Language="DocId" Value="M:System.NullReferenceException.#ctor(System.Runtime.Serialization.SerializationInfo,System.Runtime.Serialization.StreamingContext)" />
      <MemberSignature Language="VB.NET" Value="Protected Sub New (info As SerializationInfo, context As StreamingContext)" />
      <MemberSignature Language="C++ CLI" Value="protected:&#xA; NullReferenceException(System::Runtime::Serialization::SerializationInfo ^ info, System::Runtime::Serialization::StreamingContext context);" />
      <MemberSignature Language="F#" Value="new NullReferenceException : System.Runtime.Serialization.SerializationInfo * System.Runtime.Serialization.StreamingContext -&gt; NullReferenceException" Usage="new System.NullReferenceException (info, context)" />
      <MemberType>Constructor</MemberType>
      <AssemblyInfo>
        <AssemblyName>mscorlib</AssemblyName>
        <AssemblyVersion>1.0.5000.0</AssemblyVersion>
        <AssemblyVersion>2.0.0.0</AssemblyVersion>
        <AssemblyVersion>2.0.5.0</AssemblyVersion>
        <AssemblyVersion>4.0.0.0</AssemblyVersion>
      </AssemblyInfo>
      <AssemblyInfo>
        <AssemblyName>netstandard</AssemblyName>
        <AssemblyVersion>2.0.0.0</AssemblyVersion>
        <AssemblyVersion>2.1.0.0</AssemblyVersion>
      </AssemblyInfo>
      <AssemblyInfo>
        <AssemblyName>System.Runtime</AssemblyName>
        <AssemblyVersion>4.2.0.0</AssemblyVersion>
        <AssemblyVersion>4.2.1.0</AssemblyVersion>
      </AssemblyInfo>
      <Attributes>
        <Attribute FrameworkAlternate="netframework-4.0">
          <AttributeName>System.Security.SecuritySafeCritical</AttributeName>
        </Attribute>
      </Attributes>
      <Parameters>
        <Parameter Name="info" Type="System.Runtime.Serialization.SerializationInfo" Index="0" FrameworkAlternate="netcore-2.0;netcore-2.1;netcore-2.2;netcore-3.0;netframework-1.1;netframework-2.0;netframework-3.0;netframework-3.5;netframework-4.0;netframework-4.5;netframework-4.5.1;netframework-4.5.2;netframework-4.6;netframework-4.6.1;netframework-4.6.2;netframework-4.7;netframework-4.7.1;netframework-4.7.2;netframework-4.8;netstandard-2.0;xamarinandroid-7.1;xamarinios-10.8;xamarinmac-3.0;netstandard-2.1" />
        <Parameter Name="context" Type="System.Runtime.Serialization.StreamingContext" Index="1" FrameworkAlternate="netcore-2.0;netcore-2.1;netcore-2.2;netcore-3.0;netframework-1.1;netframework-2.0;netframework-3.0;netframework-3.5;netframework-4.0;netframework-4.5;netframework-4.5.1;netframework-4.5.2;netframework-4.6;netframework-4.6.1;netframework-4.6.2;netframework-4.7;netframework-4.7.1;netframework-4.7.2;netframework-4.8;netstandard-2.0;xamarinandroid-7.1;xamarinios-10.8;xamarinmac-3.0;netstandard-2.1" />
      </Parameters>
      <Docs>
        <param name="info"><span data-ttu-id="f05e3-180">Объект, содержащий сериализованные данные объекта.</span><span class="sxs-lookup"><span data-stu-id="f05e3-180">The object that holds the serialized object data.</span></span></param>
        <param name="context"><span data-ttu-id="f05e3-181">Контекстные сведения об источнике или назначении.</span><span class="sxs-lookup"><span data-stu-id="f05e3-181">The contextual information about the source or destination.</span></span></param>
        <summary><span data-ttu-id="f05e3-182">Инициализирует новый экземпляр класса <see cref="T:System.NullReferenceException" /> с сериализованными данными.</span><span class="sxs-lookup"><span data-stu-id="f05e3-182">Initializes a new instance of the <see cref="T:System.NullReferenceException" /> class with serialized data.</span></span></summary>
        <remarks>
          <format type="text/markdown"><![CDATA[  
  
## Remarks  
 <span data-ttu-id="f05e3-183">Данный конструктор вызывается в процессе десериализации для воссоздания объекта исключения, переданного в потоке.</span><span class="sxs-lookup"><span data-stu-id="f05e3-183">This constructor is called during deserialization to reconstitute the exception object transmitted over a stream.</span></span> <span data-ttu-id="f05e3-184">Дополнительные сведения см. в разделе [сериализация XML и SOAP](~/docs/standard/serialization/xml-and-soap-serialization.md).</span><span class="sxs-lookup"><span data-stu-id="f05e3-184">For more information, see [XML and SOAP Serialization](~/docs/standard/serialization/xml-and-soap-serialization.md).</span></span>  
  
 ]]></format>
        </remarks>
        <related type="Article" href="~/docs/standard/serialization/xml-and-soap-serialization.md"><span data-ttu-id="f05e3-185">Сериализация XML и SOAP</span><span class="sxs-lookup"><span data-stu-id="f05e3-185">XML and SOAP Serialization</span></span></related>
      </Docs>
    </Member>
    <Member MemberName=".ctor">
      <MemberSignature Language="C#" Value="public NullReferenceException (string message, Exception innerException);" />
      <MemberSignature Language="ILAsm" Value=".method public hidebysig specialname rtspecialname instance void .ctor(string message, class System.Exception innerException) cil managed" />
      <MemberSignature Language="DocId" Value="M:System.NullReferenceException.#ctor(System.String,System.Exception)" />
      <MemberSignature Language="VB.NET" Value="Public Sub New (message As String, innerException As Exception)" />
      <MemberSignature Language="C++ CLI" Value="public:&#xA; NullReferenceException(System::String ^ message, Exception ^ innerException);" />
      <MemberSignature Language="F#" Value="new NullReferenceException : string * Exception -&gt; NullReferenceException" Usage="new System.NullReferenceException (message, innerException)" />
      <MemberType>Constructor</MemberType>
      <AssemblyInfo>
        <AssemblyName>System.Runtime</AssemblyName>
        <AssemblyVersion>4.0.0.0</AssemblyVersion>
        <AssemblyVersion>4.0.10.0</AssemblyVersion>
        <AssemblyVersion>4.0.20.0</AssemblyVersion>
        <AssemblyVersion>4.1.0.0</AssemblyVersion>
        <AssemblyVersion>4.2.0.0</AssemblyVersion>
        <AssemblyVersion>4.2.1.0</AssemblyVersion>
      </AssemblyInfo>
      <AssemblyInfo>
        <AssemblyName>mscorlib</AssemblyName>
        <AssemblyVersion>1.0.5000.0</AssemblyVersion>
        <AssemblyVersion>2.0.0.0</AssemblyVersion>
        <AssemblyVersion>2.0.5.0</AssemblyVersion>
        <AssemblyVersion>4.0.0.0</AssemblyVersion>
      </AssemblyInfo>
      <AssemblyInfo>
        <AssemblyName>netstandard</AssemblyName>
        <AssemblyVersion>2.0.0.0</AssemblyVersion>
        <AssemblyVersion>2.1.0.0</AssemblyVersion>
      </AssemblyInfo>
      <Parameters>
        <Parameter Name="message" Type="System.String" />
        <Parameter Name="innerException" Type="System.Exception" />
      </Parameters>
      <Docs>
        <param name="message"><span data-ttu-id="f05e3-186">Сообщение об ошибке, указывающее причину создания исключения.</span><span class="sxs-lookup"><span data-stu-id="f05e3-186">The error message that explains the reason for the exception.</span></span></param>
        <param name="innerException"><span data-ttu-id="f05e3-187">Исключение, которое является причиной текущего исключения.</span><span class="sxs-lookup"><span data-stu-id="f05e3-187">The exception that is the cause of the current exception.</span></span> <span data-ttu-id="f05e3-188">Если значение параметра <paramref name="innerException" /> не равно <see langword="null" />, текущее исключение сгенерировано в блоке <see langword="catch" />, обрабатывающем внутреннее исключение.</span><span class="sxs-lookup"><span data-stu-id="f05e3-188">If the <paramref name="innerException" /> parameter is not <see langword="null" />, the current exception is raised in a <see langword="catch" /> block that handles the inner exception.</span></span></param>
        <summary><span data-ttu-id="f05e3-189">Инициализирует новый экземпляр класса <see cref="T:System.NullReferenceException" /> указанным сообщением об ошибке и ссылкой на внутреннее исключение, вызвавшее данное исключение.</span><span class="sxs-lookup"><span data-stu-id="f05e3-189">Initializes a new instance of the <see cref="T:System.NullReferenceException" /> class with a specified error message and a reference to the inner exception that is the cause of this exception.</span></span></summary>
        <remarks>
          <format type="text/markdown"><![CDATA[  
  
## Remarks  
 <span data-ttu-id="f05e3-190">Исключение, созданное как прямой результат предыдущего исключения, должно содержать в свойстве <xref:System.Exception.InnerException%2A> ссылку на предыдущее исключение.</span><span class="sxs-lookup"><span data-stu-id="f05e3-190">An exception that is thrown as a direct result of a previous exception should include a reference to the previous exception in the <xref:System.Exception.InnerException%2A> property.</span></span> <span data-ttu-id="f05e3-191">Свойство <xref:System.Exception.InnerException%2A> возвращает значение, переданное конструктору, или значение `null`, если свойство <xref:System.Exception.InnerException%2A> не передает конструктору значение внутреннего исключения.</span><span class="sxs-lookup"><span data-stu-id="f05e3-191">The <xref:System.Exception.InnerException%2A> property returns the same value that is passed into the constructor, or `null` if the <xref:System.Exception.InnerException%2A> property does not supply the inner exception value to the constructor.</span></span>  
  
 <span data-ttu-id="f05e3-192">В следующей таблице представлены исходные значения свойств экземпляра класса <xref:System.NullReferenceException>.</span><span class="sxs-lookup"><span data-stu-id="f05e3-192">The following table shows the initial property values for an instance of <xref:System.NullReferenceException>.</span></span>  
  
|<span data-ttu-id="f05e3-193">Свойство.</span><span class="sxs-lookup"><span data-stu-id="f05e3-193">Property</span></span>|<span data-ttu-id="f05e3-194">Значение</span><span class="sxs-lookup"><span data-stu-id="f05e3-194">Value</span></span>|  
|--------------|-----------|  
|<xref:System.Exception.InnerException%2A>|<span data-ttu-id="f05e3-195">Ссылка на внутреннее исключение.</span><span class="sxs-lookup"><span data-stu-id="f05e3-195">The inner exception reference.</span></span>|  
|<xref:System.Exception.Message%2A>|<span data-ttu-id="f05e3-196">Строка сообщения об ошибке.</span><span class="sxs-lookup"><span data-stu-id="f05e3-196">The error message string.</span></span>|  
  
 ]]></format>
        </remarks>
        <altmember cref="T:System.Exception" />
        <related type="Article" href="~/docs/standard/exceptions/index.md"><span data-ttu-id="f05e3-197">Обработка и создание исключений</span><span class="sxs-lookup"><span data-stu-id="f05e3-197">Handling and Throwing Exceptions</span></span></related>
      </Docs>
    </Member>
  </Members>
</Type>
